minimailgun/store.py

```diff
@@ -0,0 +1,118 @@
+
+import logging
+import pymongo
+import uuid
+from collections import Mapping
+from datetime import datetime
+from pymongo.errors import DuplicateKeyError
+from minimailgun.config import config
+
+
+log = logging.getLogger(__name__)
+
+
+class MongoStore(object):
+
+    def __init__(self, hosts=None, database=None, username=None, password=None):
+        hosts = hosts or [('127.0.0.1', None)]
+        if isinstance(hosts, Mapping):
+            hosts = hosts.items()
+        hosts = ','.join([
+            '{address}:{port}'.format(address=address, port=port or '').rstrip(':')
+            for address, port in hosts
+        ])
+
+        if bool(username) != bool(password):
+            raise ValueError('username and password are mutually required.')
+        auth = '{username}:{password}@'.format(**locals()) if username else ''
+        if not database:
+            raise ValueError('database is required.')
+
+        self.url = 'mongodb://{auth}{hosts}/{database}'.format(**locals())
+        self._database = database
+        self._clear()
+
+    @property
+    def client(self):
+        if not self._client:
```

Comment:
You're doing lazy initialization here. What will happen if multiple threads try to access this property at the same time?